pkg/apis/certificates/validation/validation_test.go

```diff
@@ -0,0 +1,237 @@
+/*
+Copyright 2020 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"crypto/ed25519"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"reflect"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	capi "k8s.io/kubernetes/pkg/apis/certificates"
+)
+
+var (
+	validObjectMeta = metav1.ObjectMeta{Name: "testcsr"}
+	validSignerName = "example.com/valid-name"
+	validUsages     = []capi.KeyUsage{capi.UsageKeyEncipherment}
+)
+
+func TestValidateCertificateSigningRequest(t *testing.T) {
+	specPath := field.NewPath("spec")
+	tests := map[string]struct {
+		csr  capi.CertificateSigningRequest
+		errs field.ErrorList
+	}{
+		"CSR with empty request data should fail": {
+			csr: capi.CertificateSigningRequest{
+				ObjectMeta: validObjectMeta,
+				Spec: capi.CertificateSigningRequestSpec{
+					Usages:     validUsages,
+					SignerName: validSignerName,
+				},
+			},
+			errs: field.ErrorList{
```

Comment:
While I think this is fine, I kind of like to compare the `string` output of the errors because that is the actual error message the user will see.  i.e. if you do a `struct` level comparison, you will not notice if your error message dumps out like a memory address or something similar.